pkg/group/manager/ldap/ldap.go

```diff
@@ -0,0 +1,396 @@
+// Copyright 2018-2020 CERN
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// In applying this license, CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+package ldap
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"fmt"
+	"strconv"
+	"strings"
+	"text/template"
+
+	"github.com/Masterminds/sprig"
+	grouppb "github.com/cs3org/go-cs3apis/cs3/identity/group/v1beta1"
+	userpb "github.com/cs3org/go-cs3apis/cs3/identity/user/v1beta1"
+	"github.com/cs3org/reva/pkg/appctx"
+	"github.com/cs3org/reva/pkg/errtypes"
+	"github.com/cs3org/reva/pkg/group"
+	"github.com/cs3org/reva/pkg/group/manager/registry"
+	"github.com/go-ldap/ldap/v3"
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+)
+
+func init() {
+	registry.Register("ldap", New)
+}
+
+type manager struct {
+	c            *config
+	groupfilter  *template.Template
+	memberfilter *template.Template
+}
+
+type config struct {
+	Hostname        string     `mapstructure:"hostname"`
+	Port            int        `mapstructure:"port"`
+	BaseDN          string     `mapstructure:"base_dn"`
+	GroupFilter     string     `mapstructure:"groupfilter"`
+	MemberFilter    string     `mapstructure:"memberfilter"`
+	AttributeFilter string     `mapstructure:"attributefilter"`
+	FindFilter      string     `mapstructure:"findfilter"`
+	BindUsername    string     `mapstructure:"bind_username"`
+	BindPassword    string     `mapstructure:"bind_password"`
+	Idp             string     `mapstructure:"idp"`
+	Schema          attributes `mapstructure:"schema"`
+}
+
+type attributes struct {
+	// DN is the distinguished name in ldap, e.g. `cn=admins,ou=groups,dc=example,dc=org`
+	DN string `mapstructure:"dn"`
+	// UID is an immutable group id, see https://docs.microsoft.com/en-us/azure/active-directory/hybrid/plan-connect-design-concepts
+	GID string `mapstructure:"gid"`
+	// CN is the group name, typically `cn`, `gid` or `samaccountname`
+	CN string `mapstructure:"cn"`
+	// Mail is the email address of a group
+	Mail string `mapstructure:"mail"`
```

Comment:
mail works for both. posixgroup requires another structural ldap schema which then might have a mail property ... should be optional anyway. for AD groups can have a mail property